### PR TITLE
Replace readItem in ProcessNumber to speed-up SolarShadingTest_ImportedShading

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------
 - Addresses #8816 
 - Replaces ```readItem``` in ```ProcessNumber``` with ```std::stod```.
 - On my Mac, this resulted in significant runtime reductions for the SolarShadingTest_ImportedShading example file.

__Runtimes for SolarShadingTest_ImportedShading__
_Before_
Run 1: 170.00 s
Run 2: 173.31 s
Run 3: 156.62 s

_After_
Run 1: 15.77 s
Run 2: 15.25 s
Run 3: 10.68 s
Run 4: 11.55 s
Run 5: 12.32 s

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jmythms** [comment]
The run time differences are amazing. I wish we knew what was causing so much improvement in Mac builds, or, what was causing Mac builds to slow down so much with the original code. The runtimes in Linux builds only had an improvement from 8s to 5s. I wonder if it is something to do with the compiler we use for Mac builds, since the code is the same.

**Comment by mitchute** [comment]
I'm not sure what the issue is. ```readItem``` [is a templated function](https://github.com/NREL/EnergyPlus/blob/5358d3eb87d1376b3484e1a38ade8274b01f0027/src/EnergyPlus/StringUtilities.hh#L79-L84), so maybe that has something to do with it. Maybe in the future we can move away from ```readItem``` for parsing strings. What did you do in the ScheduleManager refactor?

**Comment by jmythms** [comment]
The ScheduleManager refactor is focused on decreasing unnecessary [file open](https://github.com/NREL/EnergyPlus/blob/5358d3eb87d1376b3484e1a38ade8274b01f0027/src/EnergyPlus/ScheduleManager.cc#L1792), close operations where they can be avoided. I was using the original `ProcessNumber` function in it. The improvements from this PR will definitely add to the effects of the ScheduleManager refactor.

**Comment by jmythms** [comment]
Since ProcessNumber is used in three places in [ScheduleManager.cc](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/ScheduleManager.cc), any case that uses those parts of the code should see improvements from this PR:
1. Schedule:File:Shading
2. Schedule:Compact
3. Schedule:File

**Review by mjwitte — COMMENTED** [review]
@mitchute Similar timing results on Windows (just one run each, so not super accurate):
SolarShadingTest_ExternalFraction went from 214s to 113s
SolarShadingTest_ImportedShading went from 118s to 10s

This looks good so far. It would be good to have a unit test that exercises the new `catch` blocks. I've tested one path with bad values in Schedule;Compact and in the csv (I think), but a unit test that covers all the flavors of numbers this could see including bad values would be useful.

**Comment by mitchute** [comment]
Thanks, @mjwitte. Yes, I was going to do that but it looks like I didn't get it done. I'll work one up.

File: src/EnergyPlus/UtilityRoutines.cc [hunk 1]
```diff
@@ -130,23 +127,26 @@ namespace UtilityRoutines {
...
             // make FORTRAN floating point number (containing 'd' or 'D')
             // standardized by replacing 'd' or 'D' with 'e'
             std::replace_if(
                 std::begin(PString), std::end(PString), [](const char c) { return c == 'D' || c == 'd'; }, 'e');
             // then parse as a normal floating point value
-            parseFailed = !readItem(PString, rProcessNumber);
-            ErrorFlag = false;
+            try {
+                rProcessNumber = std::stod(PString, nullptr);
+                ErrorFlag = false;
```
Comment (rraustad) [inline src/EnergyPlus/UtilityRoutines.cc]: ErrorFlag = false is already set at line 128. Delete line 138.  Also, remember the discussion long ago about setting a function argument bool &ErrorFlag to false within any function. It's OK to set it true, but don't set it to false which would erase any previous error. To that end I would delete line 128 as well. @mjwitte.

I am amazed at the impact this has on execution speed!

Comment (rraustad) [reply src/EnergyPlus/UtilityRoutines.cc]: It looks like there are several places in code where this function is called that relies on ErrorFlag being true or false for each specific call. So do not delete line 128,

File: tst/EnergyPlus/unit/UtilityRoutines.unit.cc [hunk 2]
```diff
@@ -208,3 +208,52 @@ TEST_F(EnergyPlusFixture, UtilityRoutines_appendPerfLog2)
...
+    EXPECT_NEAR(UtilityRoutines::ProcessNumber(badString, expectedError), expectedVal, 1E-5);
+    EXPECT_EQ(expectedError, true);
+
+    badString = "3.14159 0";
+    expectedVal = 0.0;
+    EXPECT_NEAR(UtilityRoutines::ProcessNumber(badString, expectedError), expectedVal, 1E-5);
+    EXPECT_EQ(expectedError, true);
+
+    // out of range
+    badString = "1E5000";
```
Comment (rraustad) [inline tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: This is a bad string, not out of range. Out of range would be1.0+e5000. And shouldn't you be using EXPECT_FALSE(expectedError) or EXPECT_TRUE(expectedError)?

Also note line 259 below.

**Comment by mitchute** [comment]
OK, the unit test has been added. I marked the ```std::invalid_argument``` exception to be skipped for coverage results. I can't make it fall in there with any of the following:

"3.14159+0" - returns 3.14159
"3.14159.0" - returns 3.14159
"3.14159+-0" - returns 3.14159
"3.14159-E-0" - returns 3.14159

Maybe we should also be checking for multiple "+", "-", ".", and "E" (or similar, "d", "D", or "e"). It looks like ```std::stod``` just reads up to the first place where it finds a valid number and returns that.

File: tst/EnergyPlus/unit/UtilityRoutines.unit.cc (hunk 2 above)
Comment (mitchute) [reply tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: We're definitely falling into the ```out_of_range``` ``` catch``` block here. And yes, I should be using ```EXPECT_TRUE``` and ```EXPECT_FALSE```. I'll wait for more comments before fixing that.

Comment (rraustad) [reply tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: OK then, just wanted to be sure.

**Comment by mitchute** [comment]
Ah, I figured out the ```invalid_argument``` exception. "E3.14159" will fall in there. I'll fix that up.

Anything else?

File: tst/EnergyPlus/unit/UtilityRoutines.unit.cc (hunk 2 above)
Comment (rraustad) [reply tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: I don't see any non-valid characters in this test. As far as I can see that would wrap this up.

        if (PString.find_first_not_of(ValidNumerics) == std::string::npos) {

Comment (rraustad) [reply tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: I guess the space is non-valid so maybe your OK.

   static std::string const ValidNumerics("0123456789.+-EeDd");

**Review by rraustad — APPROVED** [review]
This all looks good with a remarkable improvement in execution speed.

File: tst/EnergyPlus/unit/UtilityRoutines.unit.cc (hunk 2 above)
Comment (mitchute) [reply tst/EnergyPlus/unit/UtilityRoutines.unit.cc:260]: We have this "É" in multiple places.

https://github.com/NREL/EnergyPlus/pull/8819/files#diff-9a44f76d9c12fb11c779204ce7954252f59a856841b113185db69cf62ad3ae56R238

**Comment by rraustad** [comment]
> OK, the unit test has been added. I marked the `std::invalid_argument` exception to be skipped for coverage results. I can't make it fall in there with any of the following:
> 
> "3.14159+0" - returns 3.14159
> "3.14159.0" - returns 3.14159
> "3.14159+-0" - returns 3.14159
> "3.14159-E-0" - returns 3.14159
> 
> Maybe we should also be checking for multiple "+", "-", ".", and "E" (or similar, "d", "D", or "e"). It looks like `std::stod` just reads up to the first place where it finds a valid number and returns that.

With this speed improvement, wouldn't it be smarter to catch that in the input processor?

**Comment by mitchute** [comment]
> wouldn't it be smarter to catch that in the input processor?

I'm honestly not sure what the right path is currently. I think we'll find that this a part of *all* input processing, regardless of whether it happens in the ```InputProcessor```. ```readItem``` is used explicitly all over the place to read in files, including the weather files. I was just looking at a file where my local profiler was saying that ```readItem``` was taking something like 20% of a 15-second simulation. We ought to take a look at seeing where we can eliminate it in other places.

**Comment by rraustad** [comment]
Just a thought, could this be moved to the input processor for numeric inputs? That certainly does not have to happen in this branch since all looks good here.

            std::replace_if(
                std::begin(PString), std::end(PString), [](const char c) { return c == 'D' || c == 'd'; }, 'e');

**Comment by jmythms** [comment]
This looks good to me. Testing done:

1. Checked with [Fortran "numerics"](https://github.com/NREL/EnergyPlus/pull/8819/files#diff-9fd10fde18fdd4ad5733be5c683249f3b21623a176c4d992a1cbb083d04e2347R130) like "2d3", the parsing worked well: Converted 2d3 to 2e3 and then to 2000.  
2. Strings like "asdf" changed the [ErrorFlag](https://github.com/NREL/EnergyPlus/pull/8819/files#diff-9fd10fde18fdd4ad5733be5c683249f3b21623a176c4d992a1cbb083d04e2347R147) to true.

File: src/EnergyPlus/UtilityRoutines.cc [hunk 3]
```diff
@@ -130,23 +127,25 @@ namespace UtilityRoutines {
...
-            ErrorFlag = false;
+            try {
+                rProcessNumber = std::stod(PString, nullptr);
+            } catch (std::invalid_argument &e) {
+                rProcessNumber = 0.0;
+                ErrorFlag = true;
+            } catch (std::out_of_range &e) {
+                rProcessNumber = 0.0;
+                ErrorFlag = true;
+            }
```
Comment (jmythms) [inline src/EnergyPlus/UtilityRoutines.cc:144]: Looks great!

File: tst/EnergyPlus/unit/UtilityRoutines.unit.cc [hunk 4]
```diff
@@ -208,3 +208,57 @@ TEST_F(EnergyPlusFixture, UtilityRoutines_appendPerfLog2)
...
+    badString = "E3.14159";
+    expectedVal = 0.0;
+    EXPECT_NEAR(UtilityRoutines::ProcessNumber(badString, expectedError), expectedVal, 1E-5);
+    EXPECT_TRUE(expectedError);
+
+    // out of range
+    badString = "1E5000";
+    expectedVal = 0.0;
+    EXPECT_NEAR(UtilityRoutines::ProcessNumber(badString, expectedError), expectedVal, 1E-5);
+    EXPECT_TRUE(expectedError);
```
Comment (jmythms) [inline tst/EnergyPlus/unit/UtilityRoutines.unit.cc:263]: I think this is comprehensive, and touched the cases I could think of.

**Review by jmythms — APPROVED** [review]

**Comment by Myoldmopar** [comment]
This is absolutely awesome!  I'll look over the code changes now but this has been reviewed by multiple folks and looks like it has a great unit test in place.  CI is really happy too.

**Review by Myoldmopar — APPROVED** [review]
OK this is ridiculous.  What an impactful change!!!

**Comment by Myoldmopar** [comment]
I'm going to build and run tests on this locally since I just merged in a couple PRs this morning.  If it passes I'll merge.  Thanks @mitchute for the work and @rraustad and @mjwitte for the reviews and testing!

**Comment by Myoldmopar** [comment]
All clean with develop merged up.  Thanks all.

**Pull request merged**